design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+4. All `PackageRevision`s that are labeled with `Package` name should be
+   deleted.
+
+*Optional Features*
+
+The following are optional features that could be implemented in a future
+iterations:
+
+- Block deletion until all CRDs are manually cleaned up.
+- Only retain a specified number of `Inactive` `PackageRevision`s. This could be
+  accomplished by adding a field to the `Package` `spec` such as
+  `revisionHistoryLimit`. The _`Package` controller_ would then clean up any
+  `PackageRevision`s that are not within the `revisionHistoryLimit`.
+
+#### PackageRevision Controller
+
+Once a `PackageRevision` has been created, the _`Package` controller_ hands
+control to the _`PackageRevision` controller_. The _`PackageRevision`
+controller_ will operate similarly to current implementation the _`Package`
+controller_, but will also handle much of the functionality currently handled by
+the existing _`Package` controller_. In total, the _`PackageRevision`_
+controller will execute the following steps:
+
+*`PackageRevision` Created with State `Pending`*
+
+1. Observe creation of a `PackageRevision`.
+2. Wait until the `spec.state` is marked as `Active`.
+3. Read the logs of the install `Job` and parsing all CRDs,
+   `InfrastructureDefinition`s, `InfrastructurePublication`s, and
+   `Composition`s. The install `Job` will be able to be located by the
+   _`PackageRevision` controller_ because it will have the same name as the
+   parent `Package`.
+   - If the hash of the image in the `initContainer` of the install `Job` does
+     not match the name of the `PackageRevision`, do not proceed.
+4. Label and add UI metadata to CRDs and composition types.
+5. Ensure that all dependent (listed in `sped.dependsOn`) CRDs are present in
+   the cluster.
+6. Create or update all parsed objects from the `Job` output, and any RBAC roles
+   that are required. These will include the current `ClusterRole`s as well as
+   those that are required to give Crossplane the ability to manage the CRDs
+   created by `InfrastructureDefinition`s and `InfrastructurePublication`s.
+7. Update CRDs produced by `InfrastructureDefinition` and
+   `InfrastructurePublication` objects with labels and UI metadata.
+8. Start the package controller as a `Deployment`.
+
+*Existing `Active` `PackageRevision` Marked as `Stopping`*
+
+1. Stop controller for package by deleting `Deployment`.
+2. Remove all ownership labels from installed CRDs, `InfrastructureDefinition`s
+   / `InfrastructurePublication`s (and their corresponding CRDs), and
+   `Composition`s.
+3. Continue to check that (1) and (2) have been accomplished until marked as
+   `Inactive`. When marked as `Inactive`, clean up install `Job`.
+
+*Existing `Inactive` `PackageRevision` Marked as `Pending`*
+
+This would occur when rollback to a previously `PackageRevision` is specified.
+
+1. Create a new install `Job` and make sure fields of the `PackageRevision`
```

Comment:
I'm getting a little confused between when the Package controller creates the install job and when the PackageRevision controller (this section) would create it.  Same with which controller reads the job output and creates those artifacts.

Do you think that Is that all sorted in your mind and this doc? 🤔

Reply:
@jbw976 this has shifted around with feedback I have received, so the `PackageRevision` section is now out of date. This is the current thinking, would love to hear your thoughts:
- `Package` controller always creates the install `Job` (or does some other form of unpacking if decided, see https://github.com/crossplane/crossplane/pull/1616#discussion_r448347779)
- `Package` controller reads objects from logs and determines if update is safe. If so, it creates / updates any CRDs / IDs / IPs / Cs.
- One of the things printed in the logs is the new `PackageRevision`. Once all objects are established, the `PackageRevision` is created.
- The `PackageRevision` has its controller `Deployment` spec and a list of the objects that were installed on its behalf. Its job is to make sure that the necessary objects exist and create the `Deployment` / keep it healthy. There is also mention that it may label the CRDs to say they are currently being used by the specific `PackageRevision`.

The model here is to make the `PackageRevision` responsible for things that are only scoped to that specific revision (i.e. the controller `Deployment`), and the `Package` responsible for things that may span across revisions as updates are made (i.e. CRDs, composition types).

Reply:
That distinction in responsibilities/focus/ownership makes sense to me